File: bindings/go/crc16_test.go
package kevy

import "testing"

// CRC16 / key_hash_slot conformance (contract §3.15 / §7).

func TestCRC16CheckVector(t *testing.T) {
	if got := crc16([]byte("123456789")); got != 0x31C3 {
		t.Fatalf("crc16 check vector: got %#x want 0x31C3", got)
	}
	if got := crc16(nil); got != 0 {
		t.Fatalf("crc16 empty: got %#x want 0", got)
	}
}

func TestHashtagRouting(t *testing.T) {
	// {tag} keys route by the tag only, so these must share a slot.
	if KeyHashSlot([]byte("{user1000}.following")) != KeyHashSlot([]byte("{user1000}.followers")) {
		t.Fatal("hashtag keys did not co-locate")
	}
	// Empty {} hashes the whole key.
	if KeyHashSlot([]byte("foo{}bar")) != KeyHashSlot([]byte("foo{}bar")) {
		t.Fatal("deterministic")
	}
	// Known Redis slot value (canonical example from the Redis docs).
	if s := KeyHashSlot([]byte("foo")); s != 12182 {
		t.Fatalf("slot(foo)=%d want 12182", s)
	}
	// The tag span equals hashing the tag contents directly.
	if KeyHashSlot([]byte("{user1000}.following")) != crc16([]byte("user1000"))&0x3FFF {
		t.Fatal("hashtag span mismatch")
	}
}

File: bindings/java/src/test/java/jp/golia/kevy/Crc16Test.java
// CRC16 + Redis-cluster key→slot mapping (client-contract §3.15 / §7).
package jp.golia.kevy;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

class Crc16Test {
    @Test void checkVector() {
        assertEquals(0x31C3, Crc16.crc16(Bytes.of("123456789")));
    }

    @Test void hashtagExtraction() {
        // {tag} routes by the tag only, so these share a slot.
        assertEquals(Crc16.keyHashSlot(Bytes.of("{user1000}.following")),
            Crc16.keyHashSlot(Bytes.of("{user1000}.followers")));
        // an empty {} span falls back to hashing the whole key.
        assertNotEquals(Crc16.keyHashSlot(Bytes.of("foo{}bar")), Crc16.keyHashSlot(Bytes.of("{}")));
    }

    @Test void slotInRange() {
        int slot = Crc16.keyHashSlot(Bytes.of("anykey"));
        assertTrue(slot >= 0 && slot < 16384);
    }
}
